WatchFaceAlphaKotlin/app/src/main/java/com/example/android/wearable/alpha/AnalogWatchFaceService.kt

```diff
@@ -0,0 +1,322 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.example.android.wearable.alpha
+
+import android.content.Context
+import android.graphics.RectF
+import android.graphics.drawable.Icon
+import android.util.Log
+import android.view.SurfaceHolder
+import androidx.wear.complications.ComplicationBounds
+import androidx.wear.complications.DefaultComplicationProviderPolicy
+import androidx.wear.complications.SystemProviders
+import androidx.wear.complications.data.ComplicationType
+import androidx.wear.watchface.CanvasComplicationDrawable
+import androidx.wear.watchface.CanvasType
+import androidx.wear.watchface.Complication
+import androidx.wear.watchface.ComplicationsManager
+import androidx.wear.watchface.WatchFace
+import androidx.wear.watchface.WatchFaceService
+import androidx.wear.watchface.WatchFaceType
+import androidx.wear.watchface.WatchState
+import androidx.wear.watchface.complications.rendering.ComplicationDrawable
+import androidx.wear.watchface.style.Layer
+import androidx.wear.watchface.style.UserStyleRepository
+import androidx.wear.watchface.style.UserStyleSchema
+import androidx.wear.watchface.style.UserStyleSetting
+import com.example.android.wearable.alpha.data.db.WatchFaceColorStyleEntity
+import com.example.android.wearable.alpha.viewmodels.AnalogWatchFaceViewModel
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.MainScope
+import kotlinx.coroutines.cancel
+import kotlinx.coroutines.launch
+
+/**
+ * Handles much of the boilerplate needed to implement a watch face (minus rendering code; see
+ * [AnalogWatchCanvasRenderer]) including the complications and settings (styles user can change on
+ * the watch face).
+ */
+class AnalogWatchFaceService : WatchFaceService() {
+    // Used to launch coroutines (non-blocking way to pull watch face color styles data [only
+    // needed on the initial load of the user color style options]).
+    private val scope: CoroutineScope = MainScope()
+
+    private lateinit var analogWatchFaceViewModel: AnalogWatchFaceViewModel
+
+    private var colorStylesList: List<WatchFaceColorStyleEntity>? = null
+    private var complicationStyleDrawableId = R.drawable.complication_white_style
+
+    override fun onCreate() {
+        super.onCreate()
+
+        analogWatchFaceViewModel =
+            AnalogWatchFaceViewModel((application as MainApplication).repository)
+
+        scope.launch {
+            colorStylesList =
+                analogWatchFaceViewModel.getAllWatchFaceColorStyles()
+        }
+    }
```

Comment:
If createUserStyles or createComplications are slow (it sounded like it might be slow on first install), you might consider starting a background thread here and scheduling their execution.  You'd have to block for the result of that in createWatchFace().  I don't think blocking in createWatchFace for short period say <100ms is a big deal.

Reply:
I might see if init works better

Reply:
NB I had a look into what was involved in making createWatchFace a suspend function and it was considerably easier than expected so we are going to change this which means in a subsequent version of the library you'll be able to avoid blocking entirely.

Reply:
That is awesome, thank you!

I will integrate that in when it is released in the upcoming alpha.